extensions/hawkbit-extension-artifact-repository-s3/src/main/java/org/eclipse/hawkbit/artifact/repository/S3Repository.java

```diff
@@ -0,0 +1,211 @@
+    @SuppressWarnings("squid:S2070")
+    public DbArtifact store(final InputStream content, final String filename, final String contentType,
+            final DbArtifactHash hash) {
+        final MessageDigest mdSHA1;
+        final MessageDigest mdMD5;
+        try {
+            mdSHA1 = MessageDigest.getInstance("SHA1");
+            mdMD5 = MessageDigest.getInstance("MD5");
+        } catch (final NoSuchAlgorithmException e) {
+            throw new ArtifactStoreException(e.getMessage(), e);
+        }
+
+        LOG.debug("Creating temporary file to store the inputstream to it");
+        final File file = createTempFile();
+
+        String sha1Hash16;
+        String md5Hash16;
+        LOG.debug("Calculating sha1 and md5 hashes");
+        try (final DigestOutputStream outputstream = openFileOutputStream(file, mdSHA1, mdMD5)) {
+            ByteStreams.copy(content, outputstream);
+            outputstream.flush();
+            sha1Hash16 = BaseEncoding.base16().lowerCase().encode(mdSHA1.digest());
+            md5Hash16 = BaseEncoding.base16().lowerCase().encode(mdMD5.digest());
+        } catch (final IOException e) {
+            throw new ArtifactStoreException(e.getMessage(), e);
+        }
+
+        return store(sha1Hash16, md5Hash16, contentType, file, hash);
+    }
+
+    private DbArtifact store(final String sha1Hash16, final String mdMD5Hash16, final String contentType,
+            final File file, final DbArtifactHash hash) {
+        final S3Artifact s3Artifact = new S3Artifact(amazonS3, s3Properties, sha1Hash16);
+        checkHashes(s3Artifact, hash);
+        LOG.info("Storing file {} with length {} to AWS S3 bucket as SHA1 {}", file.getName(), file.length(),
+                s3Properties.getBucketName(), sha1Hash16);
+        try (final InputStream inputStream = new BufferedInputStream(new FileInputStream(file),
+                RequestClientOptions.DEFAULT_STREAM_BUFFER_SIZE)) {
+
+            s3Artifact.setContentType(contentType);
+            s3Artifact.setArtifactId(sha1Hash16);
+            s3Artifact.setSize(file.length());
+            s3Artifact.setContentType(contentType);
+            s3Artifact.setHashes(new DbArtifactHash(sha1Hash16, mdMD5Hash16));
+
+            final ObjectMetadata objectMetadata = new ObjectMetadata();
+            final String mdMD5Hash64 = BaseEncoding.base64()
+                    .encode(BaseEncoding.base16().lowerCase().decode(mdMD5Hash16));
+            objectMetadata.setContentMD5(mdMD5Hash64);
+            objectMetadata.setContentType(contentType);
+            objectMetadata.setContentLength(file.length());
+            objectMetadata.setHeader("x-amz-meta-md5chksum", mdMD5Hash64);
+            if (s3Properties.isServerSideEncryption()) {
+                objectMetadata.setHeader(Headers.SERVER_SIDE_ENCRYPTION,
+                        s3Properties.getServerSideEncryptionAlgorithm());
+            }
+
+            if (exists(sha1Hash16)) {
```

Comment:
If the hash exists you do not need the `ObjectMetadata` so move its creation after this check